### PR TITLE
Avoid serializing datetime

### PR DESCRIPTION
**Pull request by adw1n**
Based on https://github.com/celery/celery/pull/5119, fixes https://github.com/celery/kombu/issues/1037

msgpack is one of the supported serializers: http://docs.celeryproject.org/en/master/userguide/calling.html#serializers and it doesn't support datetime objects serialization.

1. AsyncResult.date_done is a string, datetime object or None depending on the serializer and result backend. E.g.:
* None - amqp and rpc result backends because store_result doesn't set date_done
* string - json serializer + key-value result backend
* datetime - pickle/yaml serializer + key-value result backend
* datetime - database backends

I suggest we normalize it and make AsyncResult.date_done always be a datetime object or None. __Let me know if this is considered a breaking change.__

2. https://github.com/celery/celery/issues/4948 is about storing the time in local time with the timezone info, for now let's keep the old behaviour and use utc. It's not as simple as replacing datetime.datetime.utcnow() cals with self.app.now() because database backends use not timezone aware datetime type. Ideally you'd want to be able to change app's timezone settings
and for AsyncResult.date_done to still be correct for results saved before the configuration changes.

3. mongo is the only result backend that stores the date_done for groups, e.g.:
```
db.celery_groupmeta.find()
{ "_id" : "9be9a8aa-a265-45d8-9797-15e260c3325e", "result" : BinData(0,"ldkkNzE1OWVlZWMtOTVlOS00OGE1LTkzMzYtMjdhYjY4Y2FjZThl2SQ3NmY2YWQzNy1hOTUwLTRhNmMtYmRkNy03MWZmNGI4N2VhOGHZJGFjZDljMTE2LTk0NTktNDFlZC05ODMwLWEzZGQ4ZmQ3NDI2M9kkYzQ1YWZlODktOWU1YS00ZTE0LWE5OTEtNDlkMjgwNzY3OTRh2SQyOTc4Njg1ZS1jZDNlLTQ4ZTEtODNkNy02NmU4MjcyNGJjMzE="), "date_done" : ISODate("2019-06-17T13:53:29.190Z") }
```
GroupResult has no date_done property and when date_done is loaded using restore() only children are returned. So we don't have to do anything.

Btw. docs say about GroupResult.resore:
> Restore previously saved group result.

this is misleading because restore returns a list of children.

4. In key-value backends we check for status before setting done https://github.com/celery/celery/blob/2c029c4f33d48c1951cc97628655a04024d50a0d/celery/backends/base.py#L687 We should probably apply the same logic to other places where we set `date_done` (like mongodb, cassandra, database backends).

### CODE REVIEW

**Comment by georgepsarakis** [comment]
From a brief overview, your solution would work. Backwards compatibility could be an issue here, so I think it needs some more thought.

msgpack does offer an [API for serializing other data types]((https://github.com/msgpack/msgpack-python/#packingunpacking-of-custom-data-type), so I think it may be preferable to have this functionality available in [Kombu serialization](https://github.com/celery/kombu/blob/ee4e7bd29bf47e5b0d1259fcb5d83a55c7f8bf74/kombu/serialization.py#L357), since it could be required for other types as well.

@adw1n what do you think?

**Comment by adw1n** [comment]
> msgpack does offer an API for serializing other data types

You introduce a new reserved key (or a combination of reserved keys that mean a datetime object - '\_\_datetime\_\_' + 'as_str') and the object_hook callable is called for every dict in the result, not just the date_done. Both shouldn't be a problem, I think, unless there are millions of dicts in the result.

I'm happy with either solution.

**Comment by auvipy** [comment]
Please check the travis errors

**Comment by adw1n** [comment]
> object_hook callable is called for every dict in the result

scratch that, you can use ext_hook instead which doesn't have this drawback.

I think there is one more thing to consider. Kombu lets you register your own serializers. If you enforce that the user defined serializer must serialize datetime objects that could also be a breaking change or at lest an inconvenience for people that want to plug in their own serializers.

File: requirements/default.txt [hunk 1]
```diff
@@ -2,3 +2,5 @@ pytz>dev
 billiard>=3.6.0,<4.0
 kombu>=4.6.3,<5.0
 vine==1.3.0
+-r extras/yaml.txt
```
[the remark this answers is not in the document]
Comment (adw1n) [reply requirements/default.txt]: fixed

**Pull request merged**